port/android/tcpadapter.c

```diff
@@ -428,6 +436,16 @@ oc_tcp_receive_message(ip_context_t *dev, fd_set *fds, oc_message_t *message)
     }
   } while (total_length > message->length);
 
+  if ((session->endpoint.flags & SECURED) == 0) {
+    // validate message message before processing it
+    if (coap_tcp_parse_message(packet, message->data, message->length, true) !=
+        COAP_NO_ERROR) {
+      OC_ERR("coap_tcp_parse_message error");
+      free_tcp_session(session);
+      return ADAPTER_STATUS_ERROR;
+    }
+  }
+
   memcpy(&message->endpoint, &session->endpoint, sizeof(oc_endpoint_t));
 #ifdef OC_SECURITY
```

Comment:
If you move lines 450-454 before line 418 then you remove the endpoint argument from the `oc_tcp_is_valid_header` function and just check the `encrypted` member on the `message` argument. 
Same with the check on line `439`.

Reply:
moved